File: python/examples/09_export_csv.py
"""Persist the current snapshot to funding_<version>.csv in the working directory."""

from __future__ import annotations

import csv
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from kairo_funding import FundingClient, cli_entry  # noqa: E402

COLUMNS = (
    "exchange",
    "base",
    "funding_rate",
    "next_funding_time_ms",
    "funding_interval_hours",
    "event_time_ms",
)


def main() -> int:
    with FundingClient() as client:
        snapshot = client.get_snapshot()
    version = snapshot.get("version")
    rows = FundingClient.parse_compact_rows(snapshot.get("data", []))
    out_path = Path.cwd() / f"funding_{version}.csv"
    with out_path.open("w", newline="", encoding="utf-8") as f:
        writer = csv.writer(f, quoting=csv.QUOTE_MINIMAL)
        writer.writerow(COLUMNS)
        for row in rows:
            writer.writerow([row[c] for c in COLUMNS])
    print(f"wrote {len(rows)} rows to {out_path}")
    return 0


if __name__ == "__main__":
    cli_entry(main)

File: typescript/src/examples/09_export_csv.ts
/**
 * 09_export_csv: persist the current funding snapshot to
 * `funding_<version>.csv` in the current working directory.
 */

import { createWriteStream } from "node:fs";
import { resolve } from "node:path";
import { FundingClient } from "../client.js";
import { cliEntry } from "../runner.js";

const HEADER =
  "exchange,base,funding_rate,next_funding_time_ms,funding_interval_hours,event_time_ms";

async function main(): Promise<void> {
  const client = new FundingClient();
  const snap = await client.getSnapshot();

  const path = resolve(process.cwd(), `funding_${snap.version}.csv`);
  const stream = createWriteStream(path, { encoding: "utf8" });

  await new Promise<void>((res, rej) => {
    stream.on("error", rej);
    stream.write(HEADER + "\n");
    for (const row of snap.data) {
      stream.write(
        [
          row.exchange,
          row.base,
          row.fundingRate,
          row.nextFundingTimeMs,
          row.fundingIntervalHours,
          row.eventTimeMs,
        ].join(",") + "\n",
      );
    }
    stream.end(() => res());
  });

  process.stdout.write(`wrote ${snap.count} rows to ${path}\n`);
}

cliEntry(main);
